crates/cacache/src/put.rs

```diff
@@ -47,6 +47,165 @@ where
+            };
+
+            offset += std::io::copy(&mut entry, &mut encoded).to_internal()? as usize;
+            let (sri, _) = entry.into_inner();
+            dest = encoded.into_inner().to_internal()?;
+
+            // Would be nice if SSRI provided "into_bytes()" possibly?
+            let hexed = sri.to_hex().1;
+            offsets.insert(hex::decode(hexed).to_internal()?, offset);
+            entry_hash.insert(path, (sri, size, mode));
+        }
+
+        std::mem::drop(entry_stream);
+
+        let index_object = bincode::serialize(&entry_hash).to_internal()?;
+        let index_size = index_object.len();
+        dest.write_all(index_size.to_be_bytes().as_ref()).to_internal()?;
+        let mut encoded = snap::write::FrameEncoder::new(dest);
+        let mut entry = SSRIStream {
+            inner: std::io::Cursor::new(index_object),
+            builder: ssri::IntegrityOpts::new().algorithm(ssri::Algorithm::Sha256)
+        };
+
+        std::io::copy(&mut entry, &mut encoded).to_internal()?;
+        let (index_sri, _) = entry.into_inner();
+
+        dest = encoded.into_inner().to_internal()?;
+        let (packfile_sri, mut output_cursor) = dest.into_inner();
+
+        let offsets = offsets.into_iter().collect::<Vec<_>>();
+        let mut fanout = [0u8; 2048];
+        let mut last_index = 0;
+
+        // the fanout is a table of 256 entries, keyed positionally.
+        // the index of this fanout represents the first byte of a given
+        // integrity hash. it is mapped to an offset in a stream of hashes+offsets
+        // that follows the packfile.
+        for i in 0..255 {
+            'inner:
+            for (integrity_bytes, offset) in &offsets[last_index..] {
+                if integrity_bytes[0] != i {
+                    break 'inner;
+                }
+
+                last_index += 1;
+            }
+
+            &mut fanout[i as usize * 8..].copy_from_slice((last_index as u64).to_be_bytes().as_ref());
+        }
+
+        // write a fanout, then write the objects
+        let mut output_packidx = std::io::Cursor::new(Vec::with_capacity(fanout.len() * 8 + offsets.len() * 40)); // 32 sri bytes + 8 offset bytes
+        output_packidx.write_all(&fanout).to_internal()?;
+        for (integrity_bytes, offset) in &offsets {
+            output_packidx.write_all(integrity_bytes).to_internal()?;
+            output_packidx.write_all(offset.to_be_bytes().as_ref()).to_internal()?;
+        }
+
+        let (_, packfile_sri_hex) = packfile_sri.to_hex();
+        pb.push("packed");
+        pb.push(format!("{}.idx", packfile_sri_hex));
+        output_packidx.seek(std::io::SeekFrom::Start(0)).to_internal()?;
+        persist_cursor(output_packidx, &pb)?;
+        pb.pop();
+        pb.push(format!("{}.pack", packfile_sri_hex));
+        output_cursor.seek(std::io::SeekFrom::Start(0)).to_internal()?;
+        persist_cursor(output_cursor, &pb)?;
+
+
+        Ok(packfile_sri)
+    }).await
```

Comment:
I still have an open question about how to indicate that we should load packfiles for reading. Possibly a lazy_static / some amount of interior-to-the-module mutability is in order? Another approach might be to write the directory index as a "special" kind of object into cacache.